### ISSUE TITLE
Anomaly "Uncaught exception Not_found." Please report at http://coq.inria.fr/bugs/.

### ISSUE THREAD

**Issue by coqbot** [coq/coq#5761]
Note: the issue was created automatically with bugzilla2github tool

Original bug ID: BZ#5761
From: @JasonGross
Reported version: 8.8+alpha
CC:   @herbelin, @psteckler

**Comment by psteckler** [comment]
I think I've found the offending `Not_found`.

That exception is raised during a call to `EConstr.lookup_rel`, which is passed the result of a call to  `Context.Rel.lookup`, which can raise a `Not_found`.

In turn, `lookup_rel` is called from `Reductionops.whd_state_gen` in the `Rel` clause for `match c0`. There's no exception handler here, and I think there should be.

The backtrace provided by `ocamldebug` is:

```
#0  Pc: 3908  Pervasives char 18849
#1  Pc: 3882904  EConstr char 27043
#2  Pc: 4050876  Reductionops char 31653
#3  Pc: 4058088  Reductionops char 39896
#4  Pc: 5088016  Redexpr char 1510
#5  Pc: 4083852  Reductionops char 23758
#6  Pc: 50868  Array char 3602
#7  Pc: 3957720  Termops char 26103
#8  Pc: 50868  Array char 3602
#9  Pc: 3958728  Termops char 25885
#10  Pc: 50868  Array char 3602
#11  Pc: 3958728  Termops char 25885
#12  Pc: 50868  Array char 3602
#13  Pc: 3958728  Termops char 25885
#14  Pc: 50868  Array char 3602
#15  Pc: 3958728  Termops char 25885
#16  Pc: 50868  Array char 3602
#17  Pc: 3958728  Termops char 25885
#18  Pc: 3958292  Termops char 25536
#19  Pc: 50868  Array char 3602
#20  Pc: 3958728  Termops char 25885
#21  Pc: 50868  Array char 3602
#22  Pc: 3958728  Termops char 25885
#23  Pc: 5084140  Redexpr char 6706
#24  Pc: 5569228  Tactics char 26096
#25  Pc: 4012192  Proofview char 36828
```
There are other, similar match cases in `whd_state_gen` that might also need an exception handler.

Not sure who the expert is on this bit of code.

Ask me over a beer how I found this.

**Comment by herbelin** [comment]
Congratulation! So, this probably means that the wrong env is passed at some time, i.e. an env with a too short `rel_context`, or otherwise, a missing lift of de Bruijn indices somewhere. 

`Reductionops.whd_state_gen` is supposingly correct (if it were not, we would already know it), so
the culprit is probably in the callers to `whd_state_gen`. Is there a way with the debugger to find at what time appears the unbound `Rel` first in the hierarchy of calls, i.e. at what time the term is mentioning a `Rel` which is greater than the length of the `rel_context` part of the accompanyning env?

**Comment by psteckler** [comment]
I recall it's possible to build with plugins linked statically, which would allow further backtracking. @ppedrot knows the trick.

**Comment by ejgallego** [comment]
Using the standard build system it should suffice to pass the desired plugin `cma` to `coqmktop`.

**Comment by ejgallego** [comment]
which can be achieved by tweaking `STATICPLUGINS` in `Makefile.common`. Example:
```
 STATICPLUGINS:=plugins/ltac/ltac_plugin.cmo plugins/ltac/tauto_plugin.cmo
```

**Comment by herbelin** [comment]
Indeed, that's basically the trick I used too.

This allows to traverse plugins but setting breakpoing in packed files is broken. From 4.06 this is fixed, but only when setting breakpoint manually using the quaiifed module name. The Emacs key to set breakpoint is still broken.

**Comment by herbelin** [comment]
I just tried changing the `cbv beta` into `cbn beta` in `prove_to_of_sigma_goal` and it stopped failing. 

Apparently a serious problem with evar instances in `cbv`, a short version of it being:
```coq
Eval cbv beta in (fun x => _) 0.
(*  = ?y@{x:=_UNBOUND_REL_1} : ?T@{x:=O} *)
```

**Comment by psteckler** [comment]
> which can be achieved by tweaking STATICPLUGINS in Makefile.common

Even simpler: pass `-natdynlink no` to `configure`

**Comment by herbelin** [comment]
It seems so, indeed. I thought that, at some time, `-natdynlink` was only for dynlink in native compilation but apparently it changed. Maybe not a big deal even with `nat` in the name? What is strange though is that if the architecture does not support native compilation or if `-byte` is given, `-natdynlink no` will apparently have no effect on not dynlinking the plugins. Looks like there is a little inconsistency.

@psteckler: are you interested in working further on this bug?

**Comment by psteckler** [comment]
@herbelin In fact, I've been looking at your small example a bit. 

It appears that the `_UNBOUND_REL` in the `cbv` case is introduced during `norm_head`, in the `Lambda` clause. There's a call to `decompose_lam` (in term.ml), which calls `kind_of_term`, which returns a `Lambda` containing the `_UNBOUND_REL`. 

So `kind_of_term` is an alias for `Constr.kind`, which is just the identity, except that the input and return types differ, so it's a cast of some kind. Not sure I follow how that works, but that's where the things appear to go wrong.

**Comment by ejgallego** [comment]
`kind_of_term` is the identity, and was introduced to emulate private types so certain invariants can be enforced when building a `constr`. Thus, you need to look further for your bad bug.

**Comment by ejgallego** [comment]
I mean, the input to `kind_of_term` already had the error.

**Comment by psteckler** [comment]
I see, that's just where the error is revealed.

**Comment by psteckler** [comment]
If that indeed is an error.

**Comment by SkySkimmer** [comment]
If you have `t : constr = fun x : T => x` then `kind_of_term t` will be `Lambda (x,T, unbound rel 1)` because the body needs to be considered in the extended context.

In the example given by @herbelin `Eval cbv beta in (fun x => _) 0.` we will end up in the LAM/APP case of `cbv_stack_value` with `b = ?X@{x:=Rel 1}` and `args = [| 0 |]` and empty `stk`. So we go into `cbv_stack_term` with the extended `env`. At this point everything is fine.
The first step of `cbv_stack_term` is `norm_head`. We are in the Evar case with an undefined evar, so we return `VAL (0, ?X@{x:=Rel 1}), TOP` (TOP because empty stack).
Then we apply `cbv_stack_value` on the result. For `VAL/TOP` it just returns the VAL above.
At this point we are done with operations under the extended environment and have an unbound rel.

If we liken evars to applications it seems a `cbv_stack_term` is missing in the undefined evar case. The following patch adds it and makes both @herbelin's `Eval cbv ...` and the original report go through. I may have misunderstood something though (eg I had a first version passing `stack` which is definitively wrong).
``` ocaml
--- a/pretyping/cbv.ml
+++ b/pretyping/cbv.ml
@@ -290,7 +290,14 @@ let rec norm_head info env t stack =
   | Evar ev ->
       (match evar_value info.infos.i_cache ev with
           Some c -> norm_head info env c stack
-        | None -> (VAL(0, t), stack))
+        | None ->
+          let e, xs = ev in
+          let xs' = Array.map (fun x ->
+              let v = cbv_stack_term info TOP env x in
+              reify_value v)
+              xs
+          in
+          (VAL(0, mkEvar (e,xs')), stack))
 
   (* non-neutral cases *)
   | Lambda _ ->
```

**Comment by psteckler** [comment]
I can confirm the fix for @herbelin's example and the original bug report. @SkySkimmer do you want to make a PR (or I can do that)?

**Comment by psteckler** [comment]
Incidentally, the change does not fix #6016.
